Adding initial database schema

jonathanbossenger opened the pull request:
Documenting the Notification table schema, based on the properties of WPNotify_BaseNotification

jonathanbossenger committed:
Adding initial database schema

Documenting the Notification table schema, based on the properties of WPNotify_BaseNotification

jonathanbossenger commented:
Final schema to be determined, see conversation on #16

jonathanbossenger committed:
Adds title and action_link fields

raaaahman commented on docs/database-schema.md:
```diff
@@ -0,0 +1,13 @@
+* title `Notification message.` (optional)
+* message `Notification message.`
+* action_link `Correctly formatted URL to an internal or external action.`
+* status enum (READ, UNREAD) `Notification status.`
```
Beware that the `status` needs to be tied to each recipient. This leaves two possibilities:

1. The `recipients` field become `recipient` to only contain one recipient, and the notifications are duplicated for each recipient.
2. We create a relations table (let's name it `wp_sent_to`) which maps recipients to notification AND stores the `status` for each recipient-notification relation.

jonathanbossenger replied:
@raaaahman if we follow database normalization guidelines, we should probably create a separate table. However, I am concerned about the possible pushback from Core in creating a new table as part of (eventually) merging this feature. Are there any downsides with going with option 1?

raaaahman replied:
@jonathanbossenger I'd said that would be totally fine for the base implementation. The goal is for the design of the persistence layer to be modular, so crafty third party developers concerned by performance could replace it by whatever they fancy (Redis cache for example).

swissspidy commented on docs/database-schema.md:
```diff
@@ -0,0 +1,13 @@
+* timestamp `Timestamp of when the notification was triggered.`
+* recipients `Collection of notification recipients.`
+* title `Notification message.` (optional)
+* message `Notification message.`
```
How will title & message deal with internationalization?

If I send a notification to all users, it should be displayed in the user's language.

jonathanbossenger replied:
@swissspidy thanks for raising this point.

If we consider the current admin_notices implementation, the plugin or theme author is responsible for making their strings translatable. So off the top of my head, I'm thinking, instead of storing the title and message as text, it should be some unique key, that's relevant to a list of notification messages that the plugin or theme stores in the code. 

For example, if I have some translatable strings in an array in my plugin somewhere

`
$notifications = array(
	'plugin_slug_new_podcast_title' => __('Some title', 'plugin-slug'),
	'plugin_slug_message' => __('Some message', 'plugin-slug'),
);
`
Then when I create the new notification I instead store the keys `new_podcast_title` and `new_podcast_message` instead of the actual text.

Then WP Notify, when it sends/displays the message, can lookup these keys to get the translatable text, and translate it as required.

Obviously, this would need to be thought about and documented properly, but hopefully, you get what I mean. Does this seem like a workable solution to you?

swissspidy replied:
Yeah I think not storing the actual notification but just the key could work.

jonathanbossenger replied:
Updated schema to reflect new keys, added Translations document with a brief description of how keys can be used.

jonathanbossenger committed:
Added details for translations

raaaahman commented on docs/database-schema.md:
```diff
@@ -0,0 +1,13 @@
+
+### wp_notifications table
+* id (int) 11 auto_increment `ID of the notification.`
+* sender `Sender of the notification.`
```
The sender is becoming a model of its own. See #41 

So it may be a `sender_key` as well.

jonathanbossenger replied:
Left a comment on #41 to understand what the sender data will look like so that I can update this as needed.

raaaahman commented on docs/translations.md:
````diff
@@ -0,0 +1,25 @@
+For example, given the following list of translatable strings in a plugin 
+
+```
+$notifications = array( 
+    'plugin_slug_new_podcast_title' => __('Some title', 'plugin-slug'), 
+    'plugin_slug_message' => __('Some message', 'plugin-slug'), 
+);
````
WordPress i18n functions match translated strings with their source string (e.g.: 'Some title' ). We may be able to do the same to provide an easier API for plugin developers.

Note that this still holds the plugin responsible for returning the correct translation.

jonathanbossenger replied:
@raaaahman forgive me for not understanding. I'm not sure if this means things should change here?

raaaahman replied:
It would be easier for plugins to write just their string wrapped in a WordPress i18n functions, something like:

```php
$notification = $notification_factory->create( 
    array(
        'title' => __( 'Some title', 'plugin-domain' ),
        'message' => __( 'Some message', 'plugin-domain' )
    )
);
```

Not sure about how it'll work actually. :/

jonathanbossenger replied:
> Note that this still holds the plugin responsible for returning the correct translation.

I think this is the correct implementation, but it doesn't require any database schema changes to be possible.

raaaahman commented on docs/database-schema.md:
```diff
@@ -0,0 +1,13 @@
+* sender `Sender of the notification.`
+* timestamp `Timestamp of when the notification was triggered.`
+* recipients `Collection of notification recipients.`
+* title_key `Notification title key.`
+* message_key `Notification message key.`  (optional)
```
Maybe this two fields should be coupled in the Message model.

jonathanbossenger replied:
Is there an open issue discussing this model?

raaaahman replied:
Not yet.

jonathanbossenger replied:
@raaaahman if it's ok with you, let's merge this for now (as it's not set in stone) and we can iterate if things changed, based on the outcome of the message model discussion.

The main reason for this is that I can see the message data requirement will need to be both somehow stored by both the plugin and theme, but also needing to be dynamically generated as needed.

raaaahman replied:
If we choose option 1 from [this comment](https://github.com/WordPress/wp-notify/pull/21/files#r503378289), we'd still need to modify the `recipients` to be a single `recipient` (or `recipient_key`), otherwise we set the `status` to READ/UNREAD for all the recipients at the same time. :)

jonathanbossenger committed:
Updating sender

jonathanbossenger committed:
Updating database schema

recipient_key is the key that the plugin or theme stores (either in code, from the WordPress data, or it's own data) for the recipient.
When rendering/sending the notification, the plugin or theme needs to lookup the key, in order to determine who to show the notification to

jonathanbossenger commented:
@raaaahman I've updated the `recipients` to `recipient_key` as suggested, indicating that the data will be retrieved from a look-up on the key from the theme or plugin code/data. Please give this a once over, but otherwise, I think it's ready to merge.
